flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/BatchProjectableTableSourceScan.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.table.plan.nodes.dataset
+
+import java.util
+
+import org.apache.calcite.plan._
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rel.metadata.RelMetadataQuery
+import org.apache.calcite.rex.{RexInputRef, RexNode}
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.java.DataSet
+import org.apache.flink.api.table.BatchTableEnvironment
+import org.apache.flink.api.table.plan.schema.TableSourceTable
+import org.apache.flink.api.table.sources.{BatchTableSource, ProjectableTableSource}
+
+import scala.collection.JavaConverters._
+
+class BatchProjectableTableSourceScan(
+    cluster: RelOptCluster,
+    traits: RelTraitSet,
+    input: RelNode,
+    projects: util.List[_ <: RexNode],
+    projectionRowType: RelDataType,
+    table: RelOptTable)
+  extends BatchScan(cluster, traits, table, projectionRowType) {
+
+  override def copy(traitSet: RelTraitSet, inputs: util.List[RelNode]): RelNode = {
+    new BatchProjectableTableSourceScan(cluster,
+      traitSet,
+      input,
+      projects,
+      projectionRowType,
+      getTable)
+  }
+
+  override def computeSelfCost(planner: RelOptPlanner, mq: RelMetadataQuery): RelOptCost = {
+    super.computeSelfCost(planner, mq).multiplyBy(0.8)
+  }
+
+  override def translateToPlan(
+    tableEnv: BatchTableEnvironment,
+    expectedType: Option[TypeInformation[Any]]): DataSet[Any] = {
+
+    val tableSourceTable = getTable.unwrap(classOf[TableSourceTable])
+    val projectableSource = tableSourceTable.tableSource.asInstanceOf[ProjectableTableSource[_]]
+
+    val indexes = projects.asScala.map(_.asInstanceOf[RexInputRef].getIndex).toArray
```

Comment:
It will be error-prone, not all projects is `RexInputRef` (e.g. we `select(a, floor(b), c * 2, 1)`, only `a` is a `RexLiteral`) a exception will be thrown here.

Maybe we can add a check to the rule to make sure all the projects is `RexInputRef`.